### PR TITLE
makeNugetSource: Avoid infinite recursion in meta

### LINKED ISSUE #277628: godot3-mono.nugetSource.meta (and maybe others?): error: infinite recursion encountered
### Describe the bug

I was trying to fetch metadata for various packages for analysis and noticed that at least `godot3-mono.nugetSource.meta` attribute is failing with infinite recursion.

The reproducer against `nixpkgs-unstable`:

```
$ nix-instantiate --eval --expr 'with import <nixpkgs> {}; godot3-mono.nugetSource.meta' --show-trace
error:
       … while evaluating the attribute 'nugetSource.meta'

         at /nix/store/l71dakfm3vy5akarlqxqgqi7pdq1cblg-source/pkgs/build-support/dotnet/make-nuget-source/default.nix:29:5:

           28|   } // { # We need data from `$out` for `meta`, so we have to use overrides as to not hit infinite recursion.
           29|     meta = nuget-source.meta // {
             |     ^
           30|       licenses = let

       … in the left operand of the update (//) operator

         at /nix/store/l71dakfm3vy5akarlqxqgqi7pdq1cblg-source/pkgs/build-support/dotnet/make-nuget-source/default.nix:29:30:

           28|   } // { # We need data from `$out` for `meta`, so we have to use overrides as to not hit infinite recursion.
           29|     meta = nuget-source.meta // {
             |                              ^
           30|       licenses = let

       … while evaluating the attribute 'meta'

         at /nix/store/l71dakfm3vy5akarlqxqgqi7pdq1cblg-source/pkgs/build-support/dotnet/make-nuget-source/default.nix:29:5:

           28|   } // { # We need data from `$out` for `meta`, so we have to use overrides as to not hit infinite recursion.
           29|     meta = nuget-source.meta // {
             |     ^
           30|       licenses = let

       error: infinite recursion encountered

       at /nix/store/l71dakfm3vy5akarlqxqgqi7pdq1cblg-source/pkgs/build-support/dotnet/make-nuget-source/default.nix:29:5:

           28|   } // { # We need data from `$out` for `meta`, so we have to use overrides as to not hit infinite recursion.
           29|     meta = nuget-source.meta // {
             |     ^
           30|       licenses = let
```

### Expected behavior

Ideally metadata should evaluate successfully (or report an error with a `throw` function or equivalent, then `builtins.tryEval` could skip it).

### Notify maintainers

Not sure who the maintainer is. Trying recent contributors to `pkgs/build-support/dotnet/make-nuget-source/default.nix`.

@IvarWithoutBones, @raphaelr and @SuperSandro2000 

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.8, NixOS, 24.05 (Uakari), 24.05.20231229.6e62521`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.19.2`
 - channels(root): `""`
 - channels(slyfox): `""`
 - nixpkgs: `/run/current-system/sw/share/nixos-flakes/inputs/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by raphaelr**
## Description of changes

Slight improvement. A further step would be to get rid of the IFD, by adding license data to the nuget-deps.nix files. Closes #277628.

Tested with

```
$ nix-instantiate --eval --json --strict --option allow-import-from-derivation true . -A godot3-mono.nugetSource.meta | jq
{
  "available": true,
  "broken": false,
  "description": "A Nuget source with dependencies for godot3-mono",
  "insecure": false,
  "licenses": [
    {
      "deprecated": false,
      "free": true,
      "fullName": "MIT License",
      "redistributable": true,
      "shortName": "mit",
      "spdxId": "MIT",
      "url": "https://spdx.org/licenses/MIT.html"
    }
  ],
  "name": "godot3-mono-nuget-source",
  "outputsToInstall": [
    "out"
  ],
  "position": "/home/admin/src/nixpkgs/pkgs/build-support/dotnet/make-nuget-source/default.nix:27",
  "unfree": false,
  "unsupported": false
}
```

@NixOS/dotnet 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/build-support/dotnet/make-nuget-source/default.nix [hunk 1]
```diff
@@ -5,34 +5,31 @@
...
+  '';
 
-    meta.description = description;
-  } // { # We need data from `$out` for `meta`, so we have to use overrides as to not hit infinite recursion.
-    meta = nuget-source.meta // {
-      licenses = let
-        # TODO: avoid IFD
-        depLicenses = lib.splitString "\n" (builtins.readFile "${nuget-source}/share/licenses");
+  meta = {
+    inherit description;
```
Comment (SuperSandro2000) [inline pkgs/build-support/dotnet/make-nuget-source/default.nix:27]: Is it possible to somehow inherit all of meta here without introducing an infinite recursion again?

**Comment by raphaelr** [comment]
Urgh meta.licenses never worked anyway. And changing the attribute from `licenses` to `license` explodes with an infinite recursion error again.

```
error: evaluation aborted with the following error message:
  'Failed to evaluate ArchiSteamFarm-5.5.0.11-dependencies-source: «unknown-meta»
  has an invalid meta attrset:
         - key 'meta.licenses' is unrecognized; expected one of:
         ['available', 'badPlatforms', 'branch', 'broken', 'changelog', 'description', 'downloadPage', 'executables', 'homepage', 'hydraPlatforms', 'insecure', 'isBuildPythonPackage', 'isFcitxEngine', 'isGutenprint', 'isHydraChannel', 'isIbusEngine', 'knownVulnerabilities', 'license', 'longDescription', 'mainProgram', 'maintainers', 'maxSilent', 'name', 'outputsToInstall', 'pkgConfigModules', 'platforms', 'position', 'priority', 'schedulingPriority', 'sourceProvenance', 'tag', 'tests', 'timeout', 'unfree', 'unsupported', 'version']
```

I'm thinking of just removing the licenses attribute here completely. The go and rust builders don't set licenses for the dependencies derivation either (for example  `ripgrep.cargoDeps.meta.license` and `go-protobuf.goModules.meta.license` are both unset)

**Comment by mdarocha** [comment]
Yeah, I'm in favor of removing this license attribute, it didn't often provide any useful information, since most nuget packages don't report their license info.

File: pkgs/build-support/dotnet/make-nuget-source/default.nix (hunk 1 above)
Comment (raphaelr) [reply pkgs/build-support/dotnet/make-nuget-source/default.nix:27]: Can you clarify which meta attributes should be inherited from where?

Comment (SuperSandro2000) [reply pkgs/build-support/dotnet/make-nuget-source/default.nix:27]: Ideally just all that are provided to mkNugetSource :)

**Comment by raphaelr** [comment]
#278242

**Pull request closed without merging**